### PR TITLE
Support for Artifactory in bisector

### PR DESCRIPTION
**Pull request by ionela-voinescu**
This adds support for ArtifactoryService and modifies or adds to the documentation about uploading and downloading artifacts.

### CODE REVIEW

File: tools/bisector/bisector/bisector.py [hunk 1]
```diff
@@ -2558,6 +2563,105 @@ def upload(self, path):
...
+
+        :param path: path to the file to download
+        """
+
+        headers = {
+            "X-JFrog-Art-Api": self.token
+        }
+
+        response = requests.get(url, headers=headers)
+        # Raise an exception is the request failed
```
Comment (douglas-raillard-arm) [inline tools/bisector/bisector/bisector.py:2655]: Maybe we should just emit a warn() instead ? Raising an exception will prevent downloading any other link if one is dead for some reason. If that is the most convenient function from `requests`, we can simply:
```
try:
    ....
except Exception as e:
    warn(f'Fail to retrieve URL {url}: {e}')
```

File: tools/bisector/bisector/bisector.py [hunk 2]
```diff
@@ -5138,10 +5249,13 @@ def _main(argv):
 
     SHOW_TRACEBACK = args.debug
     service_hub = ServiceHub()
+
     try:
-        service_hub.register_service('upload', ArtifactorialService())
+        artifactorial = ArtifactorialService()
+        service_hub.register_service('upload', artifactorial)
+        service_hub.register_service('download', artifactorial)
```
Comment (douglas-raillard-arm) [inline tools/bisector/bisector/bisector.py:5256]: I guess we want to add ArtifactoryService here too. We could try ArtifactoryService first and if it raises an exception then we register the other one. This just means we have to check for the expected env var in `ArtifactoryService.__init__` and raise e.g. `ValueError` if they are not present.

**Review by douglas-raillard-arm — COMMENTED** [review]
Some nits but otherwise LGTM

File: tools/bisector/bisector/bisector.py [hunk 3]
```diff
@@ -2493,37 +2490,45 @@ def urlretrieve(url, path):
...
+        :param path: path to the file to upload
+        :param url: URL of the Artifactorial folder to upload to,
+                    or env var ARTIFACTORIAL_FOLDER
         """
 
-        url = self.url
-        token = self.token
+        token = self.token or os.getenv('ARTIFACTORIAL_TOKEN')
+        if not token:
+            raise ValueError("An Artifactorial token must be provided through ARTIFACTORIAL_TOKEN env var")
```
[the remark this answers is not in the document]
Comment (ionela-voinescu) [reply tools/bisector/bisector/bisector.py:2526]: I'm tempted to move this to __init__ but keep the format:

`self.token = token or os.getenv('ARTIFACTORIAL_TOKEN')`

as I think it's better for token to be used if given as an argument, and if not, we'll use ARTIFACTORIAL_TOKEN.

File: tools/bisector/bisector/bisector.py (hunk 1 above)
Comment (ionela-voinescu) [reply tools/bisector/bisector/bisector.py:2655]: We do catch these exceptions when we call the download function, as in:

```
                          try:
                              download_service.download(url=archive_path,
                                                        path=archive_dst)
                          except Exception as e:
                              error('Could not retrieve {archive_path}: {e}'.format(
                                  archive_path=archive_path,
                                  e=e
                              ))
```

I'm tempted to either leave it as is, or modify the download functions for both the ArtifactorialService and the ArtifactoryService to catch and warn inside the download function. 

Any preferences?

File: tools/bisector/bisector/bisector.py [hunk 4]
```diff
@@ -2558,6 +2563,105 @@ def upload(self, path):
...
+
+        headers = {
+            "X-Checksum-Md5": md5sum,
+            "X-JFrog-Art-Api": token
+        }
+
+        response = requests.put(dest, headers=headers, data=content)
+
+        if response.ok:
+            url = response.json().get("downloadUri")
```
[the remark this answers is not in the document]
Comment (ionela-voinescu) [reply tools/bisector/bisector/bisector.py:2625]: No, it cannot. It always returns a response object.

File: tools/bisector/bisector/bisector.py [hunk 5]
```diff
@@ -5138,10 +5251,21 @@ def _main(argv):
...
     SHOW_TRACEBACK = args.debug
     service_hub = ServiceHub()
+
+    artifacts_service = None
     try:
-        service_hub.register_service('upload', ArtifactorialService())
-    except Exception as e:
-        info('Artifactorial upload will not be available: ' + str(e))
+        artifacts_service = ArtifactoryService()
+        info('Artifactory Service was initialized.')
```
Comment (douglas-raillard-arm) [inline tools/bisector/bisector/bisector.py:5258]: Should be moved out in an `else` clause of the try/except/else statement as we only want to catch exceptions raised by `ArtifactoryService()`

File: tools/bisector/bisector/bisector.py [hunk 6]
```diff
@@ -5138,10 +5251,21 @@ def _main(argv):
...
+    except:
```
Comment (douglas-raillard-arm) [inline tools/bisector/bisector/bisector.py:5259]: bare `except:` are a very bad idea. They will catch any subclass of `BaseException`, which include things like `KeyboardInterrupt`. In this sort of context, the most general exception that should be caught is `Exception`

https://docs.python.org/3/library/exceptions.html#exception-hierarchy

File: tools/bisector/bisector/bisector.py [hunk 7]
```diff
@@ -5138,10 +5251,21 @@ def _main(argv):
...
+        try:
+            artifacts_service = ArtifactorialService()
+            info('Artifactorial Service was initialized.')
+        except:
```
Comment (douglas-raillard-arm) [inline tools/bisector/bisector/bisector.py:5263]: ditto

**Review by douglas-raillard-arm — CHANGES_REQUESTED** [review]

**Pull request merged**